cache/redis_test.go

```diff
@@ -0,0 +1,491 @@
+package cache_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/alicebob/miniredis/v2"
+	"github.com/go-redis/redis/v8"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/aquasecurity/fanal/cache"
+	"github.com/aquasecurity/fanal/types"
+)
+
+func TestRedisCache_PutArtifact(t *testing.T) {
+	type args struct {
+		artifactID     string
+		artifactConfig types.ArtifactInfo
+	}
+	tests := []struct {
+		name       string
+		setupRedis bool
+		args       args
+		wantKey    string
+		wantValue  string
+		wantErr    string
+	}{
+		{
+			name:       "happy path",
+			setupRedis: true,
+			args: args{
+				artifactID: "sha256:8652b9f0cb4c0599575e5a003f5906876e10c1ceb2ab9fe1786712dac14a50cf",
+				artifactConfig: types.ArtifactInfo{
+					SchemaVersion: 1,
+					Architecture:  "amd64",
+					Created:       time.Date(2020, 11, 14, 0, 20, 4, 0, time.UTC),
+					DockerVersion: "19.03.12",
+					OS:            "linux",
+				},
+			},
+			wantKey:   "fanal::artifact::sha256:8652b9f0cb4c0599575e5a003f5906876e10c1ceb2ab9fe1786712dac14a50cf",
+			wantValue: `{"SchemaVersion":1,"Architecture":"amd64","Created":"2020-11-14T00:20:04Z","DockerVersion":"19.03.12","OS":"linux"}`,
```

Comment:
Do we need that? In other places I saw that you're using the value from `args` to assert on expected JSON with `assert.JSONeq()`

Reply:
`json.Marshal` is used in `PutArtifact` / `PutBlob` and I'm afraid that the test passes even though we don't know what `json.Marshal` returns. If `json.Marshal` doesn't work as expected, we may not be able to test `PutArtifact` properly 

Let's say we want to test a function for rounding up.

```
func roundUp(a float64) float64 {
    return math.Floor(a)
```

But unfortunately, we mistakenly used `math.Floor`, not `math.Ceil`. The unit test might be as follows.

```
func TestRoundUp(t *testing.T) {
	input := 3.2
	got := roundUp(input)
	want := math.Floor(input) // should be 4.0, but 3.0 actually
	assert.Equal(t, want, got) // pass => want: 3.0, got: 3.0
}
```
 
The above test will pass even if there is a bug since `math.Floor` is also used in the test by mistake. The test should be

```
func TestRoundUp(t *testing.T) {
	input := 3.2
	got := roundUp(input)
	want := 4.0
	assert.Equal(t, want, got) // fail => want: 4.0, got: 3.0
}
```

The first test looks useless and I'm feeling like it's similar if we use `json.Marshal` in the unit test.

```
func (c RedisCache) PutBlob(blobID string, blobInfo types.BlobInfo) error {
        ...
	b, err := json.Marshal(blobInfo)
        ...
	if err := c.client.Set(context.TODO(), key, string(b), 0).Err(); err != nil {
        ...
}

func TestRedisCache_PutBlob(t *testing.T) {
	input := types.ArtifactInfo{SchemaVersion: 1}
	got :=c.PutArtifact(input)
	want := json.Marshal(input)
	assert.Equal(t, want, got)
```

I omitted some arguments/return values, but hopefully, it makes sense to you.
The test might pass even if `json.Marshal` always returns an empty value or something wrong as we don't expect.

So, I think the wanted value basically should be literal.

```
func TestRedisCache_PutBlob(t *testing.T) {
	input := types.ArtifactInfo{SchemaVersion: 1}
	got :=c.PutArtifact(input)
	want := `{"SchemaVersion":1}`
	assert.Equal(t, want, got)
```

We know the behavior of `json.Marshal` well, so the test probably works properly, but I feel it is basically doing the same thing as the example of `roundUp`.

But as you said, the string literal makes it difficult to maintain. I also don't want to use the string literal representing JSON. Actually, `PutBlob` had too long JSON and that is why I used `json.Marshal` in `TestRedisCache_PutBlob`. I don't think it is the best way.
Any better ideas?